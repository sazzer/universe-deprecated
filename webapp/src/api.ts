import axios, { Method, AxiosResponse } from 'axios';
import * as rfc6570 from 'rfc6570-expand';

declare global {
  interface Window {
    _UNIVERSE_CONFIG: any;
  }
}

/**
 * The shape of the request that we want to make
 */
export interface Request {
  url: string;
  urlParams?: any;
  method?: Method;
  headers?: any;
  data?: any;
}

export type Response<T> = AxiosResponse<T>;

/**
 * Shape of a Problem response from the serve
 */
export interface Problem {
  type: string,
  status: number,
}

/**
 * Response that is thrown if we get an RFC-7807 Problem back from the server
 */
export class ProblemResponse<T extends Problem> {
  readonly problem: T;
  readonly status: number;
  readonly headers: any;

  constructor(problem: T, status: number, headers: any) {
    this.problem = problem;
    this.status = status;
    this.headers = headers;
  }
}

/**
 * Build an API requester to make API calls with
 */
export async function request<T>(request: Request): Promise<Response<T>> {
  const serviceUrl = process.env.REACT_APP_SERVICE_URL || ((window || {})._UNIVERSE_CONFIG || {}).serviceUrl;

  const template = rfc6570.init(request.url);
  const expandedUri = template.expand(request.urlParams || {});
  try {
    return await axios.request({
      baseURL: serviceUrl,
      timeout: 20000,
      url: expandedUri,
      method: request.method,
      headers: request.headers,
      data: request.data,
    });
  } catch (e) {
    if (e.response) {
      const response = e.response as AxiosResponse<Problem>;
      if (response.headers['content-type'] === 'application/problem+json') {
        throw new ProblemResponse(response.data, response.status, response.headers);
      }
    }
    console.error('Unexpected error making API request', e);
    throw e;
  }
}
