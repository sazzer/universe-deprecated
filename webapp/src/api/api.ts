import * as rfc6570 from "rfc6570-expand";

import { Problem, ProblemResponse } from "./problem";
import axios, { AxiosResponse, Method } from "axios";

import debug from "debug";

/** The logger to use */
const LOG = debug("universe:api");

/** The error logger to use */
const ERROR_LOG = debug("universe:api:error");
debug.enable("universe:api:error");

declare global {
  interface Window {
    _UNIVERSE_CONFIG: any;
  }
}

/** The access token to use */
let _accessToken: string | undefined;

/**
 * The shape of the request that we want to make
 */
export interface Request {
  url: string;
  urlParams?: any;
  method?: Method;
  headers?: any;
  data?: any;
}

/**
 * The shape of a Response from a request
 */
export type Response<T> = AxiosResponse<T>;

/**
 * Build an API requester to make API calls with
 */
export async function request<T>(request: Request): Promise<Response<T>> {
  let serviceUrl = process.env.REACT_APP_SERVICE_URL;
  if (window && window._UNIVERSE_CONFIG && window._UNIVERSE_CONFIG.serviceUrl) {
    serviceUrl = window._UNIVERSE_CONFIG.serviceUrl;
  }

  const template = rfc6570.init(request.url);
  const expandedUri = template.expand(request.urlParams || {});
  LOG("Making request to: %s", expandedUri);

  const headers = {
    ...request.headers
  };

  if (_accessToken) {
    headers.Authorization = `Bearer ${_accessToken}`;
  }

  try {
    return await axios.request({
      baseURL: serviceUrl,
      timeout: 20000,
      url: expandedUri,
      method: request.method,
      headers: headers,
      data: request.data
    });
  } catch (e) {
    if (e.response) {
      const response = e.response as AxiosResponse<Problem>;
      if (response.headers["content-type"] === "application/problem+json") {
        LOG("Received an RFC-7807 Problem response: %o", response.data);
        throw new ProblemResponse(
          response.data,
          response.status,
          response.headers
        );
      }
    }

    ERROR_LOG("Unexpected error making API request", e);
    throw e;
  }
}

/**
 * Set the access token that we are going to use for future requests
 * @param accessToken The access token, or `undefined` to clear it
 */
export function setAccessToken(accessToken: string | undefined) {
  _accessToken = accessToken;
}
