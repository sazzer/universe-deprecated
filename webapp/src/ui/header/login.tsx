import { Link } from "react-router-dom";
import React from "react";
import { useTranslation } from "react-i18next";

/** Header link to allow the user to log in */
export const LoginLink: React.FC = () => {
  const { t } = useTranslation();

  return (
    <li className="nav-item">
      <Link to="/login" className="nav-link">
        {t("header.loginMenu.title")}
      </Link>
    </li>
  );
};
