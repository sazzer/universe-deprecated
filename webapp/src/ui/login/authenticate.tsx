import * as yup from "yup";

import { CancelButton, SubmitButton } from "../components/form/buttons";
import { ErrorMessage, FieldValues, useForm } from "react-hook-form";
import React, { useState } from "react";

import { UnexpectedError } from "../components/form/error";
import debug from "debug";
import { useTranslation } from "react-i18next";

/** The logger to use */
const LOG = debug("universe:ui:login:authenticate");

/**
 * Props for the Authenticate User page
 */
export interface AuthenticateUserPageProps {
  username: string;
  onCancel: () => void;
}

/**
 * Page for authenticating an existing user
 */
export const AuthenticateUserPage: React.FC<AuthenticateUserPageProps> = ({
  username,
  onCancel
}) => {
  const { t } = useTranslation();
  const [loading, setLoading] = useState(false);
  const [globalError, setGlobalError] = useState<string | undefined>(undefined);

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup
        .string()
        .required(
          t("login.username.errors.tag:universe,2020:validation-errors/missing")
        )
        .trim(),
      password: yup
        .string()
        .required(
          t("login.password.errors.tag:universe,2020:validation-errors/missing")
        )
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      username: username,
      password: ""
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    LOG("Submitting form: %o", data);
    setGlobalError(undefined);
    setLoading(true);
  };

  return (
    <>
      <h3>{t("login.authenticate.title")}</h3>
      <form
        onSubmit={handleSubmit(onSubmitHandler)}
        data-test="authenticateForm"
      >
        <div className="form-group" data-test="username">
          <label htmlFor="username">{t("login.username.label")}</label>
          <input
            type="text"
            className={
              errors.username ? "form-control is-invalid" : "form-control"
            }
            id="username"
            name="username"
            readOnly
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="username"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="password">
          <label htmlFor="password">{t("login.password.label")}</label>
          <input
            type="password"
            className={
              errors.password ? "form-control is-invalid" : "form-control"
            }
            id="password"
            name="password"
            autoFocus
            readOnly={loading}
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="password"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group">
          <SubmitButton loading={loading}>
            {t("login.authenticate.submit")}
          </SubmitButton>
          <CancelButton disabled={loading} onClick={onCancel}>
            {t("login.authenticate.cancel")}
          </CancelButton>
        </div>
        {globalError && <UnexpectedError message={globalError} />}
      </form>
    </>
  );
};
