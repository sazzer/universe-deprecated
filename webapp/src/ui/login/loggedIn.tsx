import React from "react";
import { useOvermind } from "../../overmind";

/**
 * React Component to render the children only when there is an active logged-in session
 */
export const LoggedIn: React.FC = ({ children }) => {
  const { state } = useOvermind();

  if (
    state.authentication.isLoggedIn ||
    process.env.NODE_ENV === "development"
  ) {
    return <>{children}</>;
  } else {
    return <div>You must be logged in for this</div>;
  }
};
