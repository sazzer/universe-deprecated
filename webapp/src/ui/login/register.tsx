import React from "react";
import { useTranslation } from "react-i18next";
import { useForm, ErrorMessage, FieldValues } from "react-hook-form";
import * as yup from "yup";
import { useOvermind } from "../../overmind";

/**
 * Render the view for the Register Form
 */
export const RegisterForm: React.FC = () => {
  const { t } = useTranslation();
  const { state, actions } = useOvermind();

  const onSubmitHandler = async (data: FieldValues) => {
    await actions.login.register({
      username: data.username,
      email: data.email,
      displayName: data.displayName,
      password: data.password
    });
  };

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup
        .string()
        .required(t("login.username.errors.required"))
        .trim(),
      email: yup
        .string()
        .email(t("login.email.errors.email"))
        .required(t("login.email.errors.required"))
        .trim(),
      displayName: yup
        .string()
        .required(t("login.displayName.errors.required"))
        .trim(),
      password: yup.string().required(t("login.password.errors.required")),
      password2: yup
        .string()
        .required(t("login.password2.errors.required"))
        .when(["password"], (password: string, schema: any) => {
          return schema.oneOf(
            [password],
            t("login.password2.errors.different")
          );
        })
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      username: state.login.username || "",
      email: "",
      displayName: "",
      password: "",
      password2: ""
    }
  });

  let errorMessage;

  return (
    <>
      <h3>{t("login.register.title")}</h3>
      <form onSubmit={handleSubmit(onSubmitHandler)} data-test="registerForm">
        <div className="form-group" data-test="username">
          <label htmlFor="username">{t("login.username.label")}</label>
          <input
            type="text"
            className={
              errors.username ? "form-control is-invalid" : "form-control"
            }
            id="username"
            name="username"
            readOnly
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="username"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="email">
          <label htmlFor="email">{t("login.email.label")}</label>
          <input
            type="text"
            className={
              errors.email ? "form-control is-invalid" : "form-control"
            }
            id="email"
            name="email"
            autoFocus
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="email"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="displayName">
          <label htmlFor="displayName">{t("login.displayName.label")}</label>
          <input
            type="text"
            className={
              errors.displayName ? "form-control is-invalid" : "form-control"
            }
            id="displayName"
            name="displayName"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="displayName"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="password">
          <label htmlFor="password">{t("login.password.label")}</label>
          <input
            type="password"
            className={
              errors.password ? "form-control is-invalid" : "form-control"
            }
            id="password"
            name="password"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="password"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="password2">
          <label htmlFor="password2">{t("login.password2.label")}</label>
          <input
            type="password"
            className={
              errors.password2 ? "form-control is-invalid" : "form-control"
            }
            id="password2"
            name="password2"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="password2"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group">
          <button
            type="submit"
            className="btn btn-primary"
            disabled={state.login.loading}
          >
            {state.login.loading && (
              <span
                className="spinner-border spinner-border-sm"
                role="status"
                aria-hidden="true"
              ></span>
            )}
            {t("login.register.submit")}
          </button>
          <button
            type="button"
            className="btn btn-link"
            disabled={state.login.loading}
            onClick={actions.login.cancelLogin}
          >
            {t("login.register.cancel")}
          </button>
        </div>
        {errorMessage}
      </form>
    </>
  );
};
