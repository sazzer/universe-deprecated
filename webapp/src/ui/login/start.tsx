import React, { useState } from 'react';
import { useTranslation } from "react-i18next";
import { useForm, ErrorMessage, FieldValues } from 'react-hook-form';
import * as yup from "yup";

/** Shape of the properties required for the Start Login Form view */
export interface StartLoginFormProps {
  onSubmit: (username: string) => Promise<void>,
}

/**
 * Render the view for the Start Login Form
 */
export const StartLoginForm: React.FC<StartLoginFormProps> = ({ onSubmit }) => {
  const { t } = useTranslation();
  const [pending, setPending] = useState(false);
  const [error, setError] = useState('');

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup.string()
        .required(t('login.username.errors.required'))
        .trim(),
    }),
    validateCriteriaMode: 'all',
    defaultValues: {
      username: ''
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    setPending(true);
    setError('');
    try {
      await onSubmit(data.username.trim());
    } catch (e) {
      setPending(false);
      setError(e.toString());
    }
  };

  let errorMessage;
  if (error) {
    errorMessage = (
      <div className="form-group">
        <div className="alert alert-danger" role="alert">
          {t('errors.unexpected', {
            message: error,
          })}
        </div>
      </div>
    );
  }
  return (
    <>
      <h3>{t('login.start.title')}</h3>
      <form onSubmit={handleSubmit(onSubmitHandler)} data-test="startLoginForm">
        <div className="form-group" data-test="username">
          <label htmlFor="username">{t('login.username.label')}</label>
          <input type="text"
            className={errors.username ? 'form-control is-invalid' : 'form-control'}
            id="username"
            name="username"
            autoFocus
            ref={register} />
          <ErrorMessage errors={errors} name="username" as={<div className="invalid-feedback" />} />
        </div>
        <div className="form-group">
          <button type="submit"
            className="btn btn-primary"
            disabled={pending}>
            {pending && <span className="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>}
            {t('login.start.submit')}
          </button>
        </div>
        {errorMessage}
      </form>
    </>
  );
};
