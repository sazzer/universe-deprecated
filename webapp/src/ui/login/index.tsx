import React, { useEffect } from "react";
import { LandingPage } from "../landing";
import { StartLoginForm } from "./start";
import { RegisterForm } from "./register";
import { AuthenticateForm } from "./authenticate";
import { useOvermind } from "../../overmind";

/**
 * Component to represent the page that is used for logging in
 */
export const LoginPage: React.FC = () => {
  const { state, actions } = useOvermind();

  useEffect(() => {
    actions.login.resetLogin();
  }, [actions.login]);

  let body;

  if (state.login.mode.current === "registering") {
    body = <RegisterForm />;
  } else if (state.login.mode.current === "authenticating") {
    body = <AuthenticateForm />;
  } else {
    body = <StartLoginForm />;
  }

  return <LandingPage>{body}</LandingPage>;
};
