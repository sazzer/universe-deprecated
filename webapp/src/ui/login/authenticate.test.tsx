import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import { Provider } from "overmind-react";
import { AuthenticateForm } from "./authenticate";
import { createTestOvermind } from "../../overmind/test";
import { OvermindMock, Config } from "overmind";
import { ValidationErrors } from "../../api/validation";
import { AuthenticationError } from "../../login/effects";

describe("AuthenticateForm", () => {
  test("Initial render", () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "authenticating"
          },
          username: "testuser"
        }
      }
    );
    const { container } = render(
      <Provider value={overmind}>
        <AuthenticateForm />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
  test("Cancelling Registration", async () => {
    const overmind = createTestOvermind(
      {},
      {
        login: {
          mode: {
            current: "authenticating"
          },
          username: "testuser"
        }
      }
    );
    const { getByText } = render(
      <Provider value={overmind}>
        <AuthenticateForm />
      </Provider>
    );
    await wait(() =>
      fireEvent.click(getByText("Cancel", { selector: "button" }))
    );
    expect(overmind.mutations).toMatchSnapshot();
  });
  describe("Submitting the form", () => {
    let authenticateEffect: jest.Mock;
    let overmind: OvermindMock<Config>;

    beforeEach(() => {
      authenticateEffect = jest.fn();

      overmind = createTestOvermind(
        {
          login: {
            api: {
              authenticateUser: authenticateEffect
            }
          }
        },
        {
          login: {
            mode: {
              current: "authenticating"
            },
            username: "testuser"
          }
        }
      );
    });
    test("Submit without entering anything", async () => {
      const { container, getByText } = render(
        <Provider value={overmind}>
          <AuthenticateForm />
        </Provider>
      );
      await wait(() =>
        fireEvent.click(getByText("Login", { selector: "button" }))
      );
      expect(container).toMatchSnapshot();
      expect(authenticateEffect).toBeCalledTimes(0);
    });
    test("Submit with whitespace password", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <AuthenticateForm />
        </Provider>
      );
      await wait(() => {
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "   " }
        });
        fireEvent.click(getByText("Login", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
      expect(authenticateEffect).toBeCalledTimes(1);
      expect(authenticateEffect).toBeCalledWith("testuser", "   ");
    });

    test("Submit successfully", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <AuthenticateForm />
        </Provider>
      );
      authenticateEffect.mockResolvedValueOnce(undefined);

      await wait(() => {
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.click(getByText("Login", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
      expect(authenticateEffect).toBeCalledTimes(1);
      expect(authenticateEffect).toBeCalledWith("testuser", "Pa55word");
    });

    test("Submit with network error", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <AuthenticateForm />
        </Provider>
      );
      authenticateEffect.mockRejectedValueOnce(new Error("Network Error"));

      await wait(() => {
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.click(getByText("Login", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
      expect(authenticateEffect).toBeCalledTimes(1);
      expect(authenticateEffect).toBeCalledWith("testuser", "Pa55word");
    });

    test("Submit with invalid password", async () => {
      const { container, getByText, getByLabelText } = render(
        <Provider value={overmind}>
          <AuthenticateForm />
        </Provider>
      );
      authenticateEffect.mockRejectedValueOnce(new AuthenticationError());

      await wait(() => {
        fireEvent.change(getByLabelText("Password"), {
          target: { value: "Pa55word" }
        });
        fireEvent.click(getByText("Login", { selector: "button" }));
      });
      expect(container).toMatchSnapshot();
      expect(authenticateEffect).toBeCalledTimes(1);
      expect(authenticateEffect).toBeCalledWith("testuser", "Pa55word");
    });
  });
});
