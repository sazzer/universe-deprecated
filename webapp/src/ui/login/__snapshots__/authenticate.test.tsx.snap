// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticateForm Cancelling Registration 1`] = `
Array [
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.username",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      null,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.error",
    "revert": [Function],
  },
  Object {
    "args": Array [
      false,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.loading",
    "revert": [Function],
  },
  Object {
    "args": Array [
      false,
    ],
    "hasChangedValue": false,
    "method": "set",
    "path": "login.loading",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
  Object {
    "args": Array [
      "initial",
    ],
    "hasChangedValue": true,
    "method": "set",
    "path": "login.mode.current",
    "revert": [Function],
  },
]
`;

exports[`AuthenticateForm Initial render 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`AuthenticateForm Submitting the form Submit successfully 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`AuthenticateForm Submitting the form Submit with invalid password 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control is-invalid"
        id="password"
        name="password"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Invalid username or password
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`AuthenticateForm Submitting the form Submit with network error 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-danger"
        role="alert"
      >
        An unexpected error occurred: Error: Network Error
      </div>
    </div>
  </form>
</div>
`;

exports[`AuthenticateForm Submitting the form Submit with whitespace password 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`AuthenticateForm Submitting the form Submit without entering anything 1`] = `
<div>
  <h3>
    Login
  </h3>
  <form
    data-test="authenticateForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="password"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        class="form-control is-invalid"
        id="password"
        name="password"
        type="password"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a password
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Login
      </button>
      <button
        class="btn btn-link"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
